### ISSUE TITLE
embark-act: Directory insert, prefix arg handling

### ISSUE THREAD

**Issue by karthink** [oantolin/embark#8]
1. Inserting a directory at `~/path/to/directory` by calling `embark-act` and then `i` inserts:

`~/path/to/directory/./` : When called from `find-file` (C-x C-f)

`~/path/to/directory/some-subdir/` : When called from `dired` (C-x d)

2. `embark-act` is not running the action when calling with prefix-arg. Trying to kill buffer and exit from `switch-to-buffer` menu:

![embark-prefix-arg](https://user-images.githubusercontent.com/8607532/82289094-d9e42200-9958-11ea-8f99-fb18c52969fb.gif)

Not sure if I'm doing something wrong. Tested both on `emacs -Q`.

**Comment by oantolin** [comment]
1. I couldn't reproduce this. For me, both from find-file and from dired, the text inserted is the same, the correct `~/path/to/directory/./` (that indeed is the first completion).

2. I see this with kill-buffer, but not with some other commands. Curious. Will investigate further.

**Comment by karthink** [comment]
Adding one more:

3. Calling shell-command `!` through `embark-act` on a file _from the completions buffer_ leads to the error

`call-shell-region: Setting current directory: Not a directory, /path/to/file` 

It works fine when called from an active minibuffer.

**Comment by oantolin** [comment]
I can't reproduce 3, and I still can't reproduce 1. The only error I do see here is 2 and I haven't figured it out. :(

**Comment by oantolin** [comment]
Oh, for 3 do you mean calling it from the completions buffer *after the minibuffer has been closed*?

**Comment by karthink** [comment]
No, sorry, I was being needlessly precise but inaccurate. I mean from the completions buffer irrespective of minibuffer status.

**Comment by oantolin** [comment]
Because I don't get any error if I call it from the completions buffer while the minibuffer is open, but I do get a different error (that the completions buffer is read only) if I call from the completions buffer when the minibuffer is no longer open. What version of Emacs are you using? I'm on 28.0.50.

**Comment by oantolin** [comment]
What's the value of `default-directory` in your completions buffer?

**Comment by karthink** [comment]
Emacs version is `27.0.90`. 
It looks like the `default-directory` is being set to whatever item the cursor is on in the completions buffer. For example, when point is on the entry `bookmarks`:

```
default-directory is a variable defined in ‘C source code’.
Its value is "~/.emacs.d.orig/bookmarks/"
Local in buffer *Completions*; global value is nil
```

**Comment by oantolin** [comment]
OK, got it. We need to fix the default-directory. And Emacs 28 knows how, I just need to track down where it is being set to the right value. I even remember seeing that when I digging about, either `minibuffer.el` or `completion.el`.

**Comment by oantolin** [comment]
OK, the `default-directory` is set by `completion-setup-function` (in `simple.el`). That function is the default value of `completion-setup-hook`. The code that sets the default directory is described as a _bad hack_. :)

```elisp
(defun completion-setup-function ()
  (let* ((mainbuf (current-buffer))
         (base-dir
          ;; FIXME: This is a bad hack.  We try to set the default-directory
          ;; in the *Completions* buffer so that the relative file names
          ;; displayed there can be treated as valid file names, independently
          ;; from the completion context.  But this suffers from many problems:
          ;; - It's not clear when the completions are file names.  With some
          ;;   completion tables (e.g. bzr revision specs), the listed
          ;;   completions can mix file names and other things.
          ;; - It doesn't pay attention to possible quoting.
          ;; - With fancy completion styles, the code below will not always
          ;;   find the right base directory.
          (if minibuffer-completing-file-name
              (file-name-as-directory
               (expand-file-name
                (buffer-substring (minibuffer-prompt-end)
                                  (- (point) (or completion-base-size 0))))))))
    (with-current-buffer standard-output
      (let ((base-size completion-base-size) ;Read before killing localvars.
            (base-position completion-base-position)
            (insert-fun completion-list-insert-choice-function))
        (completion-list-mode)
        (set (make-local-variable 'completion-base-size) base-size)
        (set (make-local-variable 'completion-base-position) base-position)
        (set (make-local-variable 'completion-list-insert-choice-function)
	     insert-fun))
      (set (make-local-variable 'completion-reference-buffer) mainbuf)
      (if base-dir (setq default-directory base-dir))
      ;; Maybe insert help string.
      (when completion-show-help
	(goto-char (point-min))
	(if (display-mouse-p)
	    (insert "Click on a completion to select it.\n"))
	(insert (substitute-command-keys
		 "In this buffer, type \\[choose-completion] to \
select the completion near point.\n\n"))))))
```

**Comment by oantolin** [comment]
Can you confirm that your `completion-setup-function` (if it exists) does not contain the "bad hack"?

**Comment by oantolin** [comment]
In the meantime, I realized you can't really use `embark` the completions buffer _after_ the minibuffer is closed: the variables needed to classify the completions are local to the minibuffer. So I added a little local variable to the completions buffer to cache the results of classification (8a8c823). 

Now you can keep the completions buffer around (maybe rename it so it doesn't get overwritten :P) long after the minibuffer is gone and still use `embark`.

**Comment by oantolin** [comment]
OK, I think I know how to fix 2 (the `C-u` stuff), but it's a little messy. And I'm actually tempted to just remove the functionality instead!

- First, it doesn't actually save you any keystrokes! It's either a `C-u` before the action or a `C-g` after the action. Now, there is normally no binding for `top-level`, but if you bind it to, say, `C-S-g`, then you'd be comparing `C-u C-u` before the action or a `C-S-g` after the action.

- Second, if you mostly use this functionality for a few functions, you can define your own action commands that include an `(abort-recursive-edit)` or `(top-level)`, like @protesilaos [did for the insert action](https://github.com/oantolin/embark/wiki/Additional-Configuration), because he prefers not to add a `C-u` or `C-u C-u` and have a single letter keybinding instead.

So between those two points I think it's best to just remove the buggy `C-u` code. What do you think @karthink?

**Comment by oantolin** [comment]
For now I removed the buggy `C-u` code. I've argued above I think we should just leave things like this, but if you want to put in back in (with a hopefully correct implementation this time!), feel free to make your case.

**Comment by oantolin** [comment]
And, for problem 1, I think I have something that renders it moot. Currently `embark` is supposed to act on the "first completion", this actually means the first completion you'd get when tab-cycling. But maybe that's not the best choice for `embark`. I think I want to change it to have an "exact match exception": if the current minibuffer contents are an acceptable completion, then use the contents as the target, even if tab-cycling would do something else.

With that change both of your `find-file` and `dired` examples would insert `~/path/to/directory/`, with no trailing `./` or other directory. That's better right?

I've implemented this change in commit a4168fc. I think this new behavior is more natural than always doing what `minibuffer-force-complete` would do (plus I can hear my mom saying: "But you have an exact match! What could be better than that? If `minibuffer-force-complete` jumped off a bridge would you do it too?")

**Comment by karthink** [comment]
2. Changes to `embark-act`: It would be good to have a mechanism to embark and exit, but it doesn't have to be a prefix argument to `embark-act`. Such a feature would make it similar to `ivy`'s actions design, where there's both an `ivy-dispatching-call` (C-M-o by default) and an `ivy-dispatching-done` (M-o by default).

I personally find myself using `ivy-dispatching-done` far more often, which would be equivalent to calling embark-act with a prefix command (until the latest commit). A common use case is calling `switch-to-buffer`, then realizing I want the buffer in some other window and calling ace-window using embark-act. In this event, I want to go to top-level immediately there's no utility to staying at the minibuffer after the call.

However, this behavior does not require an overhaul of embark-act. I tried something simple like:

```
  (defun embark-act-and-exit ()
    "Run embark action, then exit minibuffer"
    (interactive)
    (when (minibufferp)
      (embark-act)
      (top-level)))
```

And it almost works: it goes to top level while the embark transient map is still active, so the action can still be run. (I am unsure if any side effects remain.) So it would be good to modify `embark-act` slightly to accommodate something like this. If not, the current behavior is still fine with some hinting in the echo area that embark-act is still running.

Taking a step back, whether you should bundle this functionality (`embark-act-and-exit`) with `embark.el` is a tougher question:

a) Assuming the objective is to obtain more flexibility with Emacs' built in APIs (with your complementary and orthogonal suite of `icomplete-vertical`/`live-completions` and `embark`), there is merit in keeping the package as bare-bones and simple to maintain as you can, in which case you might want to omit it.

b) There is no need to copy `ivy`'s functionality function for function. My desire for `embark-act-and-exit` is colored by heavy `ivy` usage over the past year, so there is probably a different design solution here that I'm not able to think of.

**Comment by karthink** [comment]
Issue 1:
-----------------
> And, for problem 1, I think I have something that renders it moot. Currently `embark` is supposed to act on the "first completion", this actually means the first completion you'd get when tab-cycling. But maybe that's not the best choice for `embark`. I think I want to change it to have an "exact match exception": if the current minibuffer contents are an acceptable completion, then use the contents as the target, even if tab-cycling would do something else.

I tested this and it seems like a good solution, but I will try it with more commands in the next 48 hours. The function `icomplete-fido-ret` handles the specific case of `./` in `fido-mode`, and I think your solution might actually be better.

BTW, I noticed that embark-insert inserts text with properties, so you can get oddly colored text in a regular buffer that doesn't conform to existing syntax highlighting etc. You might want to insert without properties.

**Comment by karthink** [comment]
> Can you confirm that your `completion-setup-function` (if it exists) does not contain the "bad hack"?

This function is identical in emacs 27.0.1, so I doubt it's the issue.

**Comment by oantolin** [comment]
> BTW, I noticed that embark-insert inserts text with properties, so you can get oddly colored text in a regular buffer that doesn't conform to existing syntax highlighting etc. You might want to insert without properties.

Done, ditto for kill ring save.

As for the 3 problems:

1. You are testing the new exact match clause, but so far it seems OK to you.

2. You do want some action-and-exit function (I have no problem with something like that being in `embark`, if it's not insanely complicated --and it probably won't be). I'll think some more about the best way to do it some more, possibly your function + a little indicator message is good enough.

3. I still have no clue and will investigate further.

**Comment by oantolin** [comment]
Problem 3: upon further reflection I think the `file-name-as-directory` in `completion-setup-function` is a simple typo, the "bad hack" was probably intended to be with `file-name-directory`. I will report this as a bug, and in the meantime will add the following to `embark`:

```elisp
(defun embark-completions-default-directory ()
  "Guess a reasonable default directory for the completions buffer.
Meant to be added to `completion-setup-hook'."
  (when (and minibuffer-completing-file-name
             (minibufferp))
    (let ((dir (file-name-directory
                (expand-file-name
                 (buffer-substring (minibuffer-prompt-end)
                                   (- (point)
                                      (or completion-base-size 0)))))))
      (with-current-buffer standard-output
        (setq-local default-directory dir)))))
  
(add-hook 'completion-setup-hook #'embark-completions-default-directory t)
```

It's there in commit 6f4c845. Please test.

**Comment by oantolin** [comment]
When problems 1 and 3 are dealt with, I'd like to close this issue. Problem 2 lives on in issue #11.

**Comment by oantolin** [comment]
Looks like problem 2 might be solved, please test and comment over on issue #11.

**Comment by karthink** [comment]
Issue 1: The exact match exception seems to DWIM, haven't found an exception yet.
Issue 2 is mostly solved on #11.
Issue 3 is solved, default directory is being set correctly now.

**Comment by oantolin** [comment]
Excellent, thank you!
